File: src/main/java/de/schramm/royalbash/controller/service/gameevent/PlayerAttackedEvent.java
package de.schramm.royalbash.controller.service.gameevent;

import de.schramm.royalbash.controller.service.core.Game;
import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class PlayerAttackedEvent implements GameEvent {

    private final String creatureId;
    private final String ownerId;

    @Override
    public Game invoke(Game game) {
        return game
                .findPlayer(ownerId)
                .map(player-> game.
                        findCreature(creatureId)
                        .map(creature -> game
                                .combat(creature, player))
                        .orElse(game)
                ).orElse(game);
    }
}

File: src/main/java/de/schramm/royalbash/controller/service/gameevent/PlayerAttackedEvent.kt
package de.schramm.royalbash.controller.service.gameevent

import de.schramm.royalbash.controller.service.core.Game

data class PlayerAttackedEvent(
        val creatureId: String,
        val ownerId: String
) : GameEvent {

    constructor(): this(creatureId = "", ownerId = "")

    override fun invoke(game: Game): Game {
        return game
                .findPlayer(ownerId)
                .map { player -> game.findCreature(creatureId)
                        .map { creature -> game .combat(creature, player) }
                        .orElse(game) }
                .orElse(game)
    }
}
